be/src/exec/vectorized/sorting/sort_helper.h

```diff
@@ -0,0 +1,211 @@
+static inline void sort_and_tie_helper_nullable(NullableColumn* column, bool is_asc_order, bool is_null_first,
+                                                SmallPermutation& permutation, Tie& tie, std::pair<int, int> range,
+                                                bool build_tie) {
+    NullData& null_data = column->null_column_data();
+    auto null_pred = [&](const SmallPermuteItem& item) -> bool {
+        if (is_null_first) {
+            return null_data[item.index_in_chunk] == 1;
+        } else {
+            return null_data[item.index_in_chunk] != 1;
+        }
+    };
+
+#ifndef NDEBUG
+    fmt::print("nullable column tie before sort: {}\n", fmt::join(tie, ","));
+    fmt::print("nullable column before sort: {}\n", dubug_column(column, permutation));
+#endif
+
+    int range_first = range.first;
+    int range_last = 0;
+    while (range_first < range.second) {
+        range_first = find_nonzero(tie, range_first + 1);
+        if (range_first >= range.second) {
+            break;
+        }
+        range_first--;
+        range_last = find_zero(tie, range_first + 1);
+        if (range_last > range.second) {
+            break;
+        }
+        if (range_last - range_first > 1) {
+            auto pivot_iter =
+                    std::partition(permutation.begin() + range_first, permutation.begin() + range_last, null_pred);
+            int pivot_start = pivot_iter - permutation.begin();
+            int notnull_start = is_null_first ? pivot_start : range_first;
+            int notnull_end = is_null_first ? range_last : pivot_start;
+
+            if (notnull_start < notnull_end) {
+                tie[pivot_start] = 0;
+                column->data_column()->sort_and_tie(is_asc_order, is_null_first, permutation, tie,
+                                                    {notnull_start, notnull_end}, build_tie);
+            }
+        }
+
+#ifndef NDEBUG
+        fmt::print("column after iteration: [{}, {}): {}\n", range_first, range_last,
+                   dubug_column(column, permutation));
+        fmt::print("tie after iteration: [{}, {}] {}\n", range_first, range_last, fmt::join(tie, ",    "));
+#endif
+
+        range_first = range_last;
+    }
+
+#ifndef NDEBUG
+    fmt::print("nullable column tie after sort: {}\n", fmt::join(tie, ",    "));
+    fmt::print("nullable column after sort: {}\n", dubug_column(column, permutation));
+#endif
+}
+
+template <class DataComparator, class PermutationType>
+static inline void sort_and_tie_helper(Column* column, bool is_asc_order, PermutationType& permutation, Tie& tie,
+                                       DataComparator cmp, std::pair<int, int> range, bool build_tie) {
+    auto lesser = [&](auto lhs, auto rhs) { return cmp(lhs, rhs) < 0; };
+    auto greater = [&](auto lhs, auto rhs) { return cmp(lhs, rhs) > 0; };
+    auto do_sort = [&](auto begin, auto end) {
+        if (is_asc_order) {
+            ::pdqsort(false, begin, end, lesser);
```

Comment:
sort in large data set should support cancel

Reply:
done